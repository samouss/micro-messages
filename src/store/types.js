// @flow

import type { Store as ReduxStore, Dispatch as ReduxDispatch } from 'redux';
import type { MessageState, MessageAction } from '../modules/messages/types';

export type State =
  & MessageState;

export type Action =
  | MessageAction;

export type Store = ReduxStore<State, Action>;

export type GetState = () => State;

export type Thunk<A> = ((Dispatch, GetState) => Promise<void> | void) => A;

export type Dispatch =
  & ReduxDispatch<Action>
  & Thunk<Action>;
