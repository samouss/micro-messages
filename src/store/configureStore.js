// @flow

import { createStore, applyMiddleware } from 'redux';
import { createLogger } from 'redux-logger';
import thunk from 'redux-thunk';
import reducer from './rootReducer';

import type { Store } from './types';

const configureStore = (): Store => {
  const middlewares = [thunk];

  if (process.env.NODE_ENV !== 'production') {
    middlewares.push(createLogger());
  }

  const store = createStore(
    reducer,
    applyMiddleware(...middlewares),
  );

  return store;
};

export default configureStore;
