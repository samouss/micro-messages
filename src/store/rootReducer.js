// @flow

import { combineReducers } from 'redux';
import messages from 'modules/messages/reducers';

export default combineReducers({
  messages,
});
