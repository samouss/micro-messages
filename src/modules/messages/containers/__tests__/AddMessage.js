// @flow

import React from 'react';
import { shallow } from 'enzyme';
import { createMockStore } from 'test/store';
import Connect, { AddMessage } from '../AddMessage';

import type { Props } from '../AddMessage';

jest.mock('../../actions', () => ({
  postMessage: jest.fn(input => input),
}));

describe('<AddMessage />', () => {
  const createContext = () => ({
    store: createMockStore(),
  });

  const defaultProps: Props = {
    postMessage: () => Promise.resolve(),
  };

  it('expect to render the component correctly', () => {
    const props: Props = {
      ...defaultProps,
    };

    const component = shallow(
      <AddMessage
        {...props}
      />,
    );

    expect(component).toMatchSnapshot();
  });

  describe('toggleModal', () => {
    it('expect to open modal on AddMessageButton click', () => {
      const props: Props = {
        ...defaultProps,
      };

      const component = shallow(
        <AddMessage
          {...props}
        />,
      );

      component
        .find('AddMessageButton')
        .simulate('click');

      expect(component).toMatchSnapshot();
    });

    it('expect to close modal on AddMessageButton click', () => {
      const props: Props = {
        ...defaultProps,
      };

      const component = shallow(
        <AddMessage
          {...props}
        />,
      );

      // $FlowFixMe
      component.setState({ isOpen: true });

      component
        .find('AddMessageButton')
        .simulate('click');

      expect(component).toMatchSnapshot();
    });
  });

  describe('onSubmit', () => {
    it('expect to call postMessage on AddMessageForm submit', () => {
      const props: Props = {
        ...defaultProps,
        postMessage: jest.fn(() => Promise.resolve()),
      };

      const component = shallow(
        <AddMessage
          {...props}
        />,
      );

      component
        .find('AddMessageForm')
        .simulate('submit');

      expect(props.postMessage).toHaveBeenCalled();
    });

    it('expect to successfully call postMessage', () => {
      expect.assertions(2);

      const props: Props = {
        ...defaultProps,
        postMessage: jest.fn(() => Promise.resolve()),
      };

      const input = {
        body: 'Some content',
        visibility: 'public',
      };

      const toggleModal = jest.spyOn(AddMessage.prototype, 'toggleModal');

      const component = shallow(
        <AddMessage
          {...props}
        />,
      );

      // $FlowFixMe
      return component.instance().onSubmit(input).then(() => {
        expect(props.postMessage).toHaveBeenCalledWith(input);
        expect(toggleModal).toHaveBeenCalled();

        toggleModal.mockReset();
        toggleModal.mockRestore();
      });
    });

    it('expect to fail to call postMessage', () => {
      expect.assertions(2);

      const props: Props = {
        ...defaultProps,
        postMessage: jest.fn(() => Promise.reject()),
      };

      const input = {
        body: 'Some content',
        visibility: 'public',
      };

      const toggleModal = jest.spyOn(AddMessage.prototype, 'toggleModal');

      const component = shallow(
        <AddMessage
          {...props}
        />,
      );

      // $FlowFixMe
      return component.instance().onSubmit(input).then(() => {
        expect(props.postMessage).toHaveBeenCalledWith(input);
        expect(toggleModal).not.toHaveBeenCalled();

        toggleModal.mockReset();
        toggleModal.mockRestore();
      });
    });
  });

  describe('mapDisptachToProps', () => {
    it('expect to bind postMessage', () => {
      const context = createContext();

      const input = {
        body: 'Some content',
        visibility: 'public',
      };

      const component = shallow(
        <Connect />,
        { context },
      );

      // $FlowFixMe
      component.props().postMessage(input);

      expect(context.store.dispatch).toHaveBeenCalledWith(input);
    });
  });
});
