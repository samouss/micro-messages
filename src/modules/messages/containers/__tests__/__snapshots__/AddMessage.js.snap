// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddMessage /> expect to render the component correctly 1`] = `
<div>
  <Modal
    isOpen={false}
    onClose={[Function]}
  >
    <AddMessageForm
      onSubmit={[Function]}
    />
  </Modal>
  <AddMessageButton
    onClick={[Function]}
  />
</div>
`;

exports[`<AddMessage /> toggleModal expect to close modal on AddMessageButton click 1`] = `
<div>
  <Modal
    isOpen={false}
    onClose={[Function]}
  >
    <AddMessageForm
      onSubmit={[Function]}
    />
  </Modal>
  <AddMessageButton
    onClick={[Function]}
  />
</div>
`;

exports[`<AddMessage /> toggleModal expect to open modal on AddMessageButton click 1`] = `
<div>
  <Modal
    isOpen={true}
    onClose={[Function]}
  >
    <AddMessageForm
      onSubmit={[Function]}
    />
  </Modal>
  <AddMessageButton
    onClick={[Function]}
  />
</div>
`;
